src/Microsoft.Net.Http.Headers/NameValueHeaderValue.cs

```diff
@@ -0,0 +1,347 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Text;
+
+namespace Microsoft.Net.Http.Headers
+{
+    // According to the RFC, in places where a "parameter" is required, the value is mandatory
+    // (e.g. Media-Type, Accept). However, we don't introduce a dedicated type for it. So NameValueHeaderValue supports
+    // name-only values in addition to name/value pairs.
+    public class NameValueHeaderValue
+    {
+        private static readonly HttpHeaderParser<NameValueHeaderValue> SingleValueParser
+            = new GenericHeaderParser<NameValueHeaderValue>(false, GetNameValueLength);
+        internal static readonly HttpHeaderParser<NameValueHeaderValue> MultipleValueParser
+            = new GenericHeaderParser<NameValueHeaderValue>(true, GetNameValueLength);
+
+        private string _name;
+        private string _value;
+
+        private NameValueHeaderValue()
+        {
+            // Used by the parser to create a new instance of this type.
+        }
+
+        public NameValueHeaderValue(string name)
+            : this(name, null)
+        {
+        }
+
+        public NameValueHeaderValue(string name, string value)
+        {
+            CheckNameValueFormat(name, value);
+
+            _name = name;
+            _value = value;
+        }
+
+        public string Name
+        {
+            get { return _name; }
+        }
+
+        public string Value
+        {
+            get { return _value; }
+            set
+            {
+                CheckValueFormat(value);
+                _value = value;
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            Contract.Assert(_name != null);
+
+            var nameHashCode = _name.ToLowerInvariant().GetHashCode();
```

Comment:
StringComparer.OrdinalIgnoreCase.GetHashCode(string) instead of calling ToLowerInvariant.